### PR TITLE
[MIGRATE] Portal: Migrated to new api.

### PR DESCRIPTION
**Pull request by atp-odoo**
Task: https://www.odoo.com/web#id=12003&view_type=form&model=project.task&menu_id=3942&action=327
Pad: https://pad.odoo.com/p/r.q9wZCSb3Xk2xmg5n

### CODE REVIEW

File: addons/portal/models/portal.py [hunk 1]
```diff
@@ -1,14 +1,12 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from openerp.osv import fields, osv
+from openerp import fields, models
```
Comment (atp-odoo) [inline addons/portal/models/portal.py]: use from odoo import fields, models

File: addons/portal/wizard/portal_wizard.py [hunk 2]
```diff
@@ -15,43 +11,29 @@
...
-        'welcome_message': fields.text(string='Invitation Message',
-            help="This text is included in the email sent to new users of the portal."),
-    }
-
-    def _default_portal(self, cr, uid, context):
-        portal_ids = self.pool.get('res.groups').search(cr, uid, [('is_portal', '=', True)])
-        return portal_ids and portal_ids[0] or False
+    def _default_portal(self):
+        portal = self.env['res.groups'].search([('is_portal', '=', True)])
+        return portal[0] or False
```
Comment (atp-odoo) [inline addons/portal/wizard/portal_wizard.py]: return  directly self.env['res.groups'].search([('is_portal', '=', True)], limit=1)

File: addons/portal/wizard/portal_wizard.py [hunk 3]
```diff
@@ -15,43 +11,29 @@
...
+    welcome_message = fields.Text(string='Invitation Message', help="This text is included in the email sent to new users of the portal.")
 
-    def onchange_portal_id(self, cr, uid, ids, portal_id, context=None):
+    @api.multi
+    def onchange_portal_id(self, portal_id):
         # for each partner, determine corresponding portal.wizard.user records
-        res_partner = self.pool.get('res.partner')
-        partner_ids = context and context.get('active_ids') or []
+        ResPartner = self.env['res.partner']
+        partner_ids = self.env.context.get('active_ids') or []
```
Comment (atp-odoo) [inline addons/portal/wizard/portal_wizard.py]: or []  required?

File: addons/portal/wizard/portal_wizard.py [hunk 4]
```diff
@@ -66,131 +48,7 @@ def onchange_portal_id(self, cr, uid, ids, portal_id, context=None):
                     }))
         return {'value': {'user_ids': user_changes}}
 
-    def action_apply(self, cr, uid, ids, context=None):
-        wizard = self.browse(cr, uid, ids[0], context)
-        portal_user_ids = [user.id for user in wizard.user_ids]
-        self.pool.get('portal.wizard.user').action_apply(cr, uid, portal_user_ids, context)
+    @api.multi
+    def action_apply(self):
+        self[0].user_ids.action_apply()
```
Comment (atp-odoo) [inline addons/portal/wizard/portal_wizard.py]: self[0]. ? use ensure_one() and then use  self.user_ids.action_apply()

File: addons/portal/__openerp__.py [hunk 5]
```diff
@@ -22,12 +22,12 @@
...
-        'portal_view.xml',
+        'data/portal_data.xml',
+        'views/portal_views.xml',
         'wizard/portal_wizard_view.xml',
         'security/ir.model.access.csv',
     ],
-    'demo': ['portal_demo.xml'],
+    'demo': ['data/portal_demo.xml'],
     'auto_install': True,
     'installable': True,
```
Comment (atp-odoo) [inline addons/portal/__openerp__.py]: Remove version , installable. which contain default value..

File: addons/portal/data/portal_demo.xml [hunk 6]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0"?>
```
Comment (atp-odoo) [inline addons/portal/data/portal_demo.xml]: encoding="utf-8"?

File: addons/portal/wizard/portal_wizard.py [hunk 7]
```diff
@@ -15,43 +11,28 @@
...
 
-    _defaults = {
-        'portal_id': _default_portal,
-    }
+    portal_id = fields.Many2one('res.groups', domain=[('is_portal', '=', True)], required=True, string='Portal', help="The portal that users can be added in or removed from.", default=_default_portal)
+    user_ids = fields.One2many('portal.wizard.user', 'wizard_id', string='Users')
+    welcome_message = fields.Text(string='Invitation Message', help="This text is included in the email sent to new users of the portal.")
 
-    def onchange_portal_id(self, cr, uid, ids, portal_id, context=None):
+    @api.multi
```
Comment (atp-odoo) [inline addons/portal/wizard/portal_wizard.py]: why @api.multi instead of @api.onchange('portal_id')

File: addons/portal/wizard/portal_wizard.py [hunk 8]
```diff
@@ -66,131 +47,8 @@ def onchange_portal_id(self, cr, uid, ids, portal_id, context=None):
                     }))
         return {'value': {'user_ids': user_changes}}
```
Comment (atp-odoo) [inline addons/portal/wizard/portal_wizard.py]: line no: 42 to 48 should be improved. like this : 
 in_portal = portal_id in contact.user_ids[0].groups_id.ids
return self.user_ids = user_changes

File: addons/portal/wizard/portal_wizard.py [hunk 9]
```diff
@@ -66,131 +47,8 @@ def onchange_portal_id(self, cr, uid, ids, portal_id, context=None):
...
 
-    def action_apply(self, cr, uid, ids, context=None):
-        wizard = self.browse(cr, uid, ids[0], context)
-        portal_user_ids = [user.id for user in wizard.user_ids]
-        self.pool.get('portal.wizard.user').action_apply(cr, uid, portal_user_ids, context)
+    @api.multi
+    def action_apply(self):
+        self.ensure_one()
+        self.user_ids.action_apply()
```
Comment (atp-odoo) [inline addons/portal/wizard/portal_wizard.py]: :+1:

File: addons/portal/wizard/portal_wizard.py [hunk 10]
```diff
@@ -66,131 +47,8 @@ def onchange_portal_id(self, cr, uid, ids, portal_id, context=None):
...
         return {'type': 'ir.actions.act_window_close'}
```
Comment (atp-odoo) [inline addons/portal/wizard/portal_wizard.py]: is it required to return ?

File: addons/portal/wizard/portal_wizard_user.py [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+            if wizard_user.in_portal:
+                # create a user if necessary, and make sure it is in the portal group
+                if not wizard_user.user_id:
+                    user = wizard_user._create_user()
+                    wizard_user.user_id = user.id
+                if (not wizard_user.user_id.active) or (portal not in wizard_user.user_id.groups_id):
+                    wizard_user.user_id.write({'active': True, 'groups_id': [(4, portal.id)]})
+                    # prepare for the signup process
+                    wizard_user.user_id.partner_id.signup_prepare()
+                wizard_user.refresh()
```
Comment (atp-odoo) [inline addons/portal/wizard/portal_wizard_user.py]: is it required ? i think refresh is not required anymore... it's deprecated.

File: addons/portal/wizard/portal_wizard_user.py [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+            else:
+                # remove the user (if it exists) from the portal group
+                if wizard_user.user_id and (portal in wizard_user.user_id.groups_id):
+                    # if user belongs to portal only, deactivate it
+                    if len(wizard_user.user_id.groups_id) <= 1:
+                        wizard_user.user_id.write({'groups_id': [(3, portal.id)], 'active': False})
+                    else:
+                        wizard_user.user_id.write({'groups_id': [(3, portal.id)]})
+
+    def _create_user(self):
```
Comment (atp-odoo) [inline addons/portal/wizard/portal_wizard_user.py]: better to define this before action_apply function........

File: addons/portal/wizard/portal_wizard_user.py [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+        """
+        values = {
+            'email': extract_email(self.email),
+            'login': extract_email(self.email),
+            'partner_id': self.partner_id.id,
+            'groups_id': [(6, 0, [])],
+        }
+        return self.env['res.users'].with_context(noshortcut=True, no_reset_password=True).create(values)
+
+    def _send_email(self):
```
Comment (atp-odoo) [inline addons/portal/wizard/portal_wizard_user.py]: same define this method before action_apply

File: addons/portal/__openerp__.py [hunk 14]
```diff
@@ -22,12 +21,10 @@
...
     'data': [
-        'portal_data.xml',
-        'portal_view.xml',
+        'data/portal_data.xml',
+        'views/portal_views.xml',
         'wizard/portal_wizard_view.xml',
         'security/ir.model.access.csv',
     ],
-    'demo': ['portal_demo.xml'],
-    'auto_install': True,
```
Comment (atp-odoo) [inline addons/portal/__openerp__.py]: default value for auto_install is False, here required "auto_install": True............... ;)

File: addons/portal/wizard/portal_wizard_user.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+    def _create_user(self):
+        """ create a new user for wizard_user.partner_id
+            @param wizard_user: browse record of model portal.wizard.user
+            @return: browse record of model res.users
+        """
+        values = {
+            'email': extract_email(self.email),
+            'login': extract_email(self.email),
+            'partner_id': self.partner_id.id,
+            'groups_id': [(6, 0, [])],
```
Comment (atp-odoo) [inline addons/portal/wizard/portal_wizard_user.py]: is this required ?

File: addons/portal/wizard/portal_wizard.py [hunk 16]
```diff
@@ -15,182 +11,43 @@
...
-        for partner in res_partner.browse(cr, SUPERUSER_ID, partner_ids, context):
+        for partner in ResPartner.sudo().browse(partner_ids):
             for contact in (partner.child_ids or [partner]):
                 # make sure that each contact appears at most once in the list
                 if contact.id not in contact_ids:
                     contact_ids.add(contact.id)
                     in_portal = False
                     if contact.user_ids:
-                        in_portal = portal_id in [g.id for g in contact.user_ids[0].groups_id]
+                        in_portal = self.portal_id in contact.user_ids[0].groups_id.ids
```
Comment (WHPODOO) [inline addons/portal/wizard/portal_wizard.py]: If you store list of ids into variable then variable name should be with suffix of ```_id```

File: addons/portal/wizard/portal_wizard_user.py [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+        template = self.env.ref('portal.mail_template_data_portal_welcome')
+        if template:
+            template.with_context({'dbname': self.env.cr.dbname, 'portal_url': portal_url}).send_mail(self.id, force_send=True)
+        else:
+            _logger.warning("No email template found for sending email to the portal user")
+        return True
+
+    @api.multi
+    def action_apply(self):
+        error_msg = self.get_error_messages()
```
Comment (hmo-odoo) [inline addons/portal/wizard/portal_wizard_user.py]: is it possible batch operations in wizard ? if not, can we use ensure_one() ?

File: addons/portal/wizard/portal_wizard.py [hunk 18]
```diff
@@ -15,182 +11,43 @@
...
+    def onchange_portal_id(self):
         # for each partner, determine corresponding portal.wizard.user records
-        res_partner = self.pool.get('res.partner')
-        partner_ids = context and context.get('active_ids') or []
+        ResPartner = self.env['res.partner']
+        partner_ids = self.env.context.get('active_ids')
         contact_ids = set()
         user_changes = []
-        for partner in res_partner.browse(cr, SUPERUSER_ID, partner_ids, context):
+        for partner in ResPartner.sudo().browse(partner_ids):
```
Comment (xmo-odoo) [inline addons/portal/wizard/portal_wizard.py]: Why not `self.env['res.partner'].sudo()` directly?

File: addons/portal/wizard/portal_wizard.py (hunk 16 above)
Comment (xmo-odoo) [inline addons/portal/wizard/portal_wizard.py]: is the `.ids` really needed? groups_id should be a `res.groups` recordset and `self.portal_id` also.

File: addons/portal/wizard/portal_wizard_user.py [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+        error_user = []
+        ctx = dict(self.env.context or {}, active_test=False)
+        for wizard_user in self:
+            if wizard_user.in_portal and not wizard_user.user_id:
+                email = extract_email(wizard_user.email)
+                if not email:
+                    error_empty.append(wizard_user.partner_id)
+                elif email in emails and email not in error_emails:
+                    error_emails.append(wizard_user.partner_id)
+                user = res_users.sudo().with_context(ctx).search([('login', '=', email)])
```
Comment (xmo-odoo) [inline addons/portal/wizard/portal_wizard_user.py]: Why not define `res_users` as `self.env['res.users'].sudo().with_context(active_test=False)` directly?

File: addons/portal/wizard/portal_wizard_user.py [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+                if user:
+                    error_user.append(wizard_user.partner_id)
+                emails.append(email)
+
+        error_msg = []
+        if error_empty:
+            error_msg.append("%s\n- %s" % (_("Some contacts don't have a valid email: "),
+                                '\n- '.join(['%s' % (p.display_name,) for p in error_empty])))
```
Comment (xmo-odoo) [inline addons/portal/wizard/portal_wizard_user.py]: * no need for the list, a generator comprehension will work
* why the string formatting? `p.display_name` is already a string no?

File: addons/portal/wizard/portal_wizard_user.py [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+        if error_emails:
+            error_msg.append("%s\n- %s" % (_("Several contacts have the same email: "),
+                                '\n- '.join([p.email for p in error_emails])))
```
Comment (xmo-odoo) [inline addons/portal/wizard/portal_wizard_user.py]: See notes above

File: addons/portal/wizard/portal_wizard_user.py [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+        if error_user:
+            error_msg.append("%s\n- %s" % (_("Some contacts have the same email as an existing portal user:"),
+                                '\n- '.join(['%s <%s>' % (p.display_name, p.email) for p in error_user])))
```
Comment (xmo-odoo) [inline addons/portal/wizard/portal_wizard_user.py]: generator comprehension should suffice, no need for the list part

File: addons/portal/wizard/portal_wizard_user.py [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+        self.ensure_one()
+        if not self.env.user.email:
+            raise UserError(_('You must have an email address in your User Preferences to send emails.'))
+
+        # determine subject and body in the portal user's language
+        context = dict(self.env.context or {}, lang=self.user_id.lang)
+        ctx_portal_url = dict(context, signup_force_type_in_url='')
+        partner = self.partner_id
+        portal_url = partner.with_context(ctx_portal_url)._get_signup_url_for_action()[partner.id]
+        partner.with_context(context).signup_prepare()
```
Comment (xmo-odoo) [inline addons/portal/wizard/portal_wizard_user.py]: Why not just use `**` and only keyword arguments merged into current context?

File: addons/portal/wizard/portal_wizard_user.py [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+        template = self.env.ref('portal.mail_template_data_portal_welcome')
+        if template:
+            template.with_context({'dbname': self.env.cr.dbname, 'portal_url': portal_url}).send_mail(self.id, force_send=True)
```
Comment (xmo-odoo) [inline addons/portal/wizard/portal_wizard_user.py]: semantics change. Old code added keys to existing context, this one completely replaces context to only use these keys.

File: addons/portal/wizard/portal_wizard_user.py [hunk 25]
```diff
@@ -0,0 +1,123 @@
...
+        """ create a new user for wizard_user.partner_id
+        """
+        values = {
+            'email': extract_email(self.email),
+            'login': extract_email(self.email),
+            'partner_id': self.partner_id.id,
+        }
+        return self.env['res.users'].with_context(noshortcut=True, no_reset_password=True).create(values)
+
+    @api.multi
```
Comment (gpa-odoo) [inline addons/portal/wizard/portal_wizard_user.py]: Here decorator required?

File: addons/portal/wizard/portal_wizard_user.py [hunk 26]
```diff
@@ -0,0 +1,123 @@
...
+        portal_url = partner.with_context(lang=self.user_id.lang, signup_force_type_in_url='')._get_signup_url_for_action()[partner.id]
+        partner.with_context(lang=self.user_id.lang).signup_prepare()
+        template = self.env.ref('portal.mail_template_data_portal_welcome')
+        if template:
+            template.with_context(dbname=self.env.cr.dbname, portal_url=portal_url, lang=self.user_id.lang).send_mail(self.id, force_send=True)
+        else:
+            _logger.warning("No email template found for sending email to the portal user")
+        return True
+
+    def action_apply(self):
```
Comment (gpa-odoo) [inline addons/portal/wizard/portal_wizard_user.py]: decorater ?

File: addons/portal/tests/test_portal.py [hunk 27]
```diff
@@ -1,7 +1,7 @@
 # -*- coding: utf-8 -*-
 
-from openerp.addons.mail.tests.common import TestMail
-from openerp.tools.misc import mute_logger
+from odoo.addons.mail.tests.common import TestMail
+from odoo.tools.misc import mute_logger
 
 
 class test_portal(TestMail):
```
Comment (gpa-odoo) [inline addons/portal/tests/test_portal.py]: why not here class name in camelcase?

File: addons/portal/models/portal.py [hunk 28]
```diff
@@ -1,14 +1,12 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from openerp.osv import fields, osv
+from odoo import fields, models
 
 
-class portal(osv.osv):
+class portal(models.Model):
```
Comment (gpa-odoo) [inline addons/portal/models/portal.py]: why not here name in camelcase?

File: addons/portal/tests/test_portal.py (hunk 27 above)
Comment (gpa-odoo) [inline addons/portal/tests/test_portal.py]: where user of "base_url" variable ?

**Comment by jem-odoo** [comment]
Landed in master at https://github.com/odoo/odoo/commit/a9d11f1175ff36f2ad295d08626c073e1317c533

**Pull request closed without merging**